--- ../../../../gcc3.4.6/gcc-3.4.6/gcc/cpperror.c	2003-11-01 23:56:54.000000000 +0100
+++ cpperror.c	2007-12-17 23:01:09.000000000 +0100
@@ -29,8 +29,15 @@
 #include "cpphash.h"
 #include "intl.h"
 
+#include "system.h"
+#include "coretypes.h"
+#include "tm.h" 
+
 static void print_location (cpp_reader *, fileline, unsigned int);
 
+const char *throwback_file = "";
+int throwback_line = 0;
+
 /* Print the logical file location (LINE, COL) in preparation for a
    diagnostic.  Outputs the #include chain if it has changed.  A line
    of zero suppresses the include stack, and outputs the program name
@@ -59,6 +66,9 @@
       else
 	fprintf (stderr, "%s:%u:%u:", map->to_file, lin, col);
 
+      throwback_file = map->to_file;
+      throwback_line = lin;
+
       fputc (' ', stderr);
     }
 }
@@ -146,7 +156,19 @@
     }
 
   if (_cpp_begin_message (pfile, level, line, column))
-    v_message (msgid, ap);
+    {
+#ifdef ERROR_THROWBACK
+      char buffer[256];
+      va_list ap2;
+
+      va_copy (ap2, ap);
+      v_message (msgid, ap);
+      vsnprintf (buffer, sizeof(buffer), _(msgid), ap2);
+      ERROR_THROWBACK (throwback_file, throwback_line, CPP_DL_WARNING_P (level) ? "warning" : "", buffer);
+#else
+      v_message (msgid, ap);
+#endif
+    }
 
   va_end (ap);
 }
@@ -162,7 +184,19 @@
   va_start (ap, msgid);
 
   if (_cpp_begin_message (pfile, level, line, column))
-    v_message (msgid, ap);
+    {
+#ifdef ERROR_THROWBACK
+      char buffer[256];
+      va_list ap2;
+
+      va_copy (ap2, ap);
+      v_message (msgid, ap);
+      vsnprintf (buffer, sizeof(buffer), _(msgid), ap2);
+      ERROR_THROWBACK (throwback_file, throwback_line, CPP_DL_WARNING_P (level) ? "warning" : "", buffer);
+#else
+      v_message (msgid, ap);
+#endif
+    }
 
   va_end (ap);
 }
