Index: c-parse.in
===================================================================
--- c-parse.in	(revision 6787)
+++ c-parse.in	(working copy)
@@ -635,7 +635,7 @@
 	IDENTIFIER
 		{
 		  if (yychar == YYEMPTY)
-		    yychar = YYLEX;
+		    yychar = yylex ();
 		  $$ = build_external_ref ($1, yychar == '(');
 		}
 	| CONSTANT
@@ -2227,7 +2227,7 @@
 
 save_location:
 		{ if (yychar == YYEMPTY)
-		    yychar = YYLEX;
+		    yychar = yylex ();
 		  $$ = input_location; }
 	;
 
